### PR TITLE
Google Cloud Compute Guide

### PR DESCRIPTION
**Pull request by shingjo**
- Created Base Google Deploy
- Added guide for deploying to Google Compute
- Updated existing Google Run, Run Native and App Engine to use base
- Changed Guide merge logic & Category can be defined only in base metadata

### CODE REVIEW

File: buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy [hunk 1]
```diff
@@ -60,6 +60,10 @@ class GuideProjectGenerator implements AutoCloseable {
 
         mergeMetadataList(metadatas)
 
+        if (metadatas.publish && !metadatas.category) {
```
Comment (burtbeckwith) [inline buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy]: `metadatas.publish` and `metadatas.category` are lists (shorthand for `*.`) so this will always be false since `metadatas.category` won't be empty

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 2]
```diff
@@ -0,0 +1,123 @@
+common:header.adoc[]
+
+You will deploy a Micronaut application to https://cloud.google.com/compute/docs[Google Cloud Compute Engine] is a computing and hosting service.
```
Comment (burtbeckwith) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]: "is a" -> ", a"

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+You will see an output such as:
+
+[source, bash, options="nowrap"]
+----
+Created [https://www.googleapis.com/compute/v1/projects/micronaut-guides-jdbc/zones/us-east1-b/instances/guide-micronaut-jdbc-compute].
+NAME                          ZONE        MACHINE_TYPE  PREEMPTIBLE  INTERNAL_IP    EXTERNAL_IP     STATUS
+guide-micronaut-jdbc-compute  us-east1-b  e2-micro                   10.xxx.xxx.4   34.xxx.xxx.200  RUNNING
+----
+
+Make note of your external IP address, will need this later.
```
Comment (burtbeckwith) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]: "will" -> "you will" or "you'll"

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+=== Open the 8080 Port
```
Comment (burtbeckwith) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]: "Open the 8080 Port" -> "Open Port 8080"

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+[source, bash]
+----
+gcloud compute firewall-rules create allow-8080 --allow tcp:8080
+----
+
+You will see an output such as:
```
Comment (burtbeckwith) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]: remove "an"

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+----
+gcloud compute scp build/libs/micronautguide-0.1-all.jar \
+    guide-micronaut-jdbc-compute:.
+----
+
+NOTE: If you have not set up SSH key file for `gcloud`, you will be prompted to set a passphrase to set public/private RSA key pair.
+
+=== Start Micronaut Application
+
+Before we can start the application, need to access our newly created instance.
```
Comment (burtbeckwith) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]: "need" -> "we need"

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+[source,bash]
+----
+gcloud compute ssh guide-micronaut-jdbc-compute
+----
+
+Once we are connected to our instance.
```
Comment (burtbeckwith) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]: "Once" -> "Now"

File: buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy (hunk 1 above)
Comment (shingjo) [reply buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy]: No idea what I was thinking.

**Comment by sdelamo** [comment]
@shingjo please use the github project for guides. I have added the PR to the project.

File: buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy [hunk 8]
```diff
@@ -319,16 +319,16 @@ class GuideProjectGenerator implements AutoCloseable {
         merged.intro = metadata.intro ?: base.intro
         merged.authors = mergeLists(base.authors, metadata.authors)
         merged.tags = mergeLists(base.tags, metadata.tags)
-        merged.category = base.category ?: metadata.category
+        merged.category = metadata.category ?: base.category
```
Comment (sdelamo) [inline buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy]: can you move this change to a separate PR so that we can merge it as soon as possible? Thanks

Comment (sdelamo) [reply buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy]: I've moved this to a PR https://github.com/micronaut-projects/micronaut-guides/pull/765

File: guides/micronaut-google-app-engine/metadata.json [hunk 9]
```diff
@@ -1,8 +1,9 @@
 {
+  "base": "micronaut-google-deploy-base",
   "title": "Deploy to Google App Engine",
   "intro": "Learn how to deploy a Micronaut application to Google App Engine Java Flexible Environment",
   "authors": ["Sergio del Amo"],
-  "tags": ["appengine", "gcp"],
```
Comment (sdelamo) [inline guides/micronaut-google-app-engine/metadata.json]: why delete `gcp` tag? Is that not a common accronym to refer to Google Cloud Platform?

File: guides/micronaut-google-cloud-platform-cloud-run/metadata.json [hunk 10]
```diff
@@ -1,8 +1,9 @@
 {
+  "base": "micronaut-google-deploy-base",
   "title": "Deploy to Google Cloud Run",
   "intro": "Deploy a Micronaut application to Google Cloud Run - a fully managed serverless platform for containerized applications.",
   "authors": ["Sergio del Amo"],
-  "tags": ["gcp", "cloudrun"],
```
Comment (sdelamo) [inline guides/micronaut-google-cloud-platform-cloud-run/metadata.json]: why delete `gcp` tag? Is that not a common accronym to refer to Google Cloud Platform?

**Comment by shingjo** [comment]
@sdelamo [Deploy GraalVM Native Image to Google Cloud Run](https://guides.micronaut.io/latest/micronaut-graalvm-native-image-google-cloud-platform-cloud-run.html) currently does not have a groovy guide.

**Comment by sdelamo** [comment]
I've left this PR for the new guide

File: buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy [hunk 11]
```diff
@@ -75,7 +75,7 @@ class GuideProjectGenerator implements AutoCloseable {
 
         Category cat = Category.values().find {it.toString() == config.category }
         if (publish && !cat) {
-            throw new GradleException("$config.category does not exist in Category enum")
+            throw new GradleException("$configFile.parentFile.name metadata.category=$config.category does not exist in Category enum")
```
Comment (sdelamo) [inline buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy:78]: why this?

**Comment by sdelamo** [comment]
@kroegerk can you proofread this?

File: buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy (hunk 11 above)
Comment (shingjo) [reply buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy:78]: Just made the error message more expressive, which is helpful when editing multiple files. Now it tells you which project had the issue.

**Comment by graemerocher** [comment]
Can this be updated to cover deploying the native image?

**Comment by shingjo** [comment]
@graemerocher I left out native image for now, since need this guide as a reference for the GCP MySQL guide. Will circle back and add steps for native image.

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+[source, bash]
+----
+gcloud compute instances create \
+  guide-micronaut-jdbc-compute \
+  --machine-type=e2-micro \
+  --zone=us-east1-b
+----
+
+NOTE: if you've omitted to specify a default zone in the setup steps, the command will respond requesting a zone for your instance.
```
Comment (kroegerk) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]:
```suggestion
NOTE: If you've failed to specify a default zone in the setup steps, the command will respond requesting a zone for your instance.
```
Alternately,
```suggestion
NOTE: If you've omitted the default zone in the setup steps, the command will respond requesting a zone for your instance.
```

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+You will see output such as:
```
Comment (kroegerk) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]:
```suggestion
You will see the following output:
```

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+[source, bash, options="nowrap"]
+----
+Created [https://www.googleapis.com/compute/v1/projects/micronaut-guides-jdbc/zones/us-east1-b/instances/guide-micronaut-jdbc-compute].
+NAME                          ZONE        MACHINE_TYPE  PREEMPTIBLE  INTERNAL_IP    EXTERNAL_IP     STATUS
+guide-micronaut-jdbc-compute  us-east1-b  e2-micro                   10.xxx.xxx.4   34.xxx.xxx.200  RUNNING
+----
+
+Make note of your external IP address, you will need this later.
```
Comment (kroegerk) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]:
```suggestion
Make note of your external IP address; you will need this later.
```

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+=== Open Port 8080
+
+[source, bash]
+----
+gcloud compute firewall-rules create allow-8080 --allow tcp:8080
+----
+
+You will see output such as:
```
Comment (kroegerk) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]:
```suggestion
You will see the following output:
```

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+
+=== Copy the JAR to the Instance
+
+[source,bash]
+----
+gcloud compute scp build/libs/micronautguide-0.1-all.jar \
+    guide-micronaut-jdbc-compute:.
+----
+
+NOTE: If you have not set up SSH key file for `gcloud`, you will be prompted to set a passphrase to set public/private RSA key pair.
```
Comment (kroegerk) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]:
```suggestion
NOTE: If you have not set up an SSH key file for `gcloud`, you will be prompted to set a passphrase to set a public/private RSA key pair.
```
The verb "set" is used three times in this sentence. Could the last one be changed to "define"? ("set a passphrase to *define* a public/private RSA key pair")

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+Now we can run our Micronaut application.
+
+[,bash]
+----
+java -jar micronautguide-0.1-all.jar
+----
+
+=== Running Application
+
+Verify that the application is running by invoking the controller at http://[VM IP Address]:8080 in a browser or using cURL:
```
Comment (kroegerk) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc]:
```suggestion
Verify that the application is running either by invoking the controller at http://[VM IP Address]:8080 in a browser or by using cURL:
```

**Review by burtbeckwith — APPROVED** [review]

**Review by kroegerk — APPROVED** [review]

File: guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc [hunk 18]
```diff
@@ -91,8 +91,8 @@ Let's make sure it is up-to-date and install the Java JRE.
 
 [,bash]
 ----
-apt update
-sudo apt install default-jre
+sudo apt update
+apt install default-jre
```
Comment (burtbeckwith) [inline guides/micronaut-google-cloud-compute/micronaut-google-cloud-compute.adoc:95]: both commands need `sudo`

**Comment by sdelamo** [comment]
> apt install default-jre

I did not need `sudo` for the second one. If you think I missed something please add it.
